File: src/utils/color.ts
import { WingColor } from "@/bindings";

export const ALL_WING_COLORS = [
  "GrayBlue",
  "MediumBlue",
  "DarkBlue",
  "Turquoise",
  "Green",
  "OliveGreen",
  "Yellow",
  "Orange",
  "Red",
  "Coral",
  "Pink",
  "Mauve",
] as const;

export const getWingColor = (wingColor: WingColor): string => {
  switch (wingColor) {
    case "GrayBlue":
      return "#9FE0EA";
    case "MediumBlue":
      return "#0BF1FB";
    case "DarkBlue":
      return "#28B6E8";
    case "Turquoise":
      return "#00FBEE";
    case "Green":
      return "#00EE4D";
    case "OliveGreen":
      return "#C6DC52";
    case "Yellow":
      return "#FFEF3D";
    case "Orange":
      return "#FF8D3D";
    case "Red":
      return "#FF3933";
    case "Coral":
      return "#FF916B";
    case "Pink":
      return "#FDA5F6";
    case "Mauve":
      return "#9F92FA";
  }
};

File: src-tauri/src/wing/color.rs
#[repr(u8)]
#[derive(Copy, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, specta::Type)]
pub enum WingColor {
    GrayBlue = 1,
    MediumBlue = 2,
    DarkBlue = 3,
    Turquoise = 4,
    Green = 5,
    OliveGreen = 6,
    Yellow = 7,
    Orange = 8,
    Red = 9,
    Coral = 10,
    Pink = 11,
    Mauve = 12,
}

impl TryFrom<u8> for WingColor {
    type Error = ();

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            1 => Ok(WingColor::GrayBlue),
            2 => Ok(WingColor::MediumBlue),
            3 => Ok(WingColor::DarkBlue),
            4 => Ok(WingColor::Turquoise),
            5 => Ok(WingColor::Green),
            6 => Ok(WingColor::OliveGreen),
            7 => Ok(WingColor::Yellow),
            8 => Ok(WingColor::Orange),
            9 => Ok(WingColor::Red),
            10 => Ok(WingColor::Coral),
            11 => Ok(WingColor::Pink),
            12 => Ok(WingColor::Mauve),
            _ => Err(()),
        }
    }
}
